--- a/modules/proxy/mod_proxy_http.c	
+++ a/modules/proxy/mod_proxy_http.c	
@@ -701,7 +701,7 @@ int ap_proxy_http_request(apr_pool_t *p, request_rec *r,
     int force10, rv;
     apr_table_t *headers_in_copy;
 
-    header_brigade = apr_brigade_create(p, origin->bucket_alloc);
+    header_brigade = apr_brigade_create(p, bucket_alloc);
 
     /*
      * Send the HTTP/1.1 request to the remote server
@@ -1853,6 +1853,11 @@ apr_status_t ap_proxy_http_process_response(apr_pool_t * p, request_rec *r,
         }
     } while (interim_response && (interim_response < AP_MAX_INTERIM_RESPONSES));
 
+    /* We have to cleanup bb brigade, because buckets inserted to it could be
+     * created from scpool and this pool can be freed before this brigade. */
+    apr_brigade_cleanup(bb);
+
+
     /* See define of AP_MAX_INTERIM_RESPONSES for why */
     if (interim_response >= AP_MAX_INTERIM_RESPONSES) {
         return ap_proxyerror(r, HTTP_BAD_GATEWAY,
