--- a/modules/proxy/mod_proxy_connect.c	2010/12/08 19:29:34	1043627
+++ b/modules/proxy/mod_proxy_connect.c	2010/12/08 19:31:34	1043628
@@ -288,6 +288,9 @@
 
     while (1) { /* Infinite loop until error (one side closes the connection) */
         if ((rv = apr_pollset_poll(pollset, -1, &pollcnt, &signalled)) != APR_SUCCESS) {
+            if (APR_STATUS_IS_EINTR(rv)) { 
+                continue;
+            }
             apr_socket_close(sock);
             ap_log_rerror(APLOG_MARK, APLOG_ERR, rv, r, "proxy: CONNECT: error apr_poll()");
             return HTTP_INTERNAL_SERVER_ERROR;
